File: java/src/scheduler/Scheduler.java
package scheduler;

import java.util.*;

public class Scheduler {
    
    public static final String[] DAYS = {
        "Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday", "Sunday"
    };
    public static final String[] SHIFTS = {"Morning", "Afternoon", "Evening"};
    public static final int MAX_DAYS_PER_WEEK = 5;
    public static final int MIN_EMPLOYEES_PER_SHIFT = 2;
    
    public List<String> employees;
    public Map<String, Map<String, List<String>>> preferences;
    private Map<String, Map<String, List<String>>> schedule;
    private Map<String, Integer> daysWorked;
    private Map<String, Set<String>> employeeAssignedDays;
    
    public Scheduler() {
        this.employees = new ArrayList<>();
        this.preferences = new HashMap<>();
        this.schedule = new HashMap<>();
        this.daysWorked = new HashMap<>();
        this.employeeAssignedDays = new HashMap<>();
        
        // Initialize schedule structure
        for (String day : DAYS) {
            Map<String, List<String>> shiftsMap = new HashMap<>();
            for (String shift : SHIFTS) {
                shiftsMap.put(shift, new ArrayList<>());
            }
            schedule.put(day, shiftsMap);
        }
    }
    
    public void addEmployee(String name) {
        if (!employees.contains(name)) {
            employees.add(name);
            preferences.put(name, new HashMap<>());
            daysWorked.put(name, 0);
            employeeAssignedDays.put(name, new HashSet<>());
            System.out.println("Added employee: " + name);
        } else {
            System.out.println("Employee " + name + " already exists.");
        }
    }
    
    public void addPreference(String employee, String day, List<String> shifts) {
        if (!employees.contains(employee)) {
            System.out.println("Error: Employee " + employee + " not found. Add them first.");
            return;
        }
        
        if (!Arrays.asList(DAYS).contains(day)) {
            System.out.println("Error: Invalid day " + day);
            return;
        }
        
        List<String> validShifts = new ArrayList<>();
        for (String shift : shifts) {
            if (Arrays.asList(SHIFTS).contains(shift)) {
                validShifts.add(shift);
            }
        }
        
        if (validShifts.isEmpty()) {
            System.out.println("Error: No valid shifts provided");
            return;
        }
        
        preferences.get(employee).put(day, validShifts);
        System.out.println("Added preference for " + employee + " on " + day + ": " + 
                         String.join(", ", validShifts));
    }
    
    public void assignShifts() {
        System.out.println("\n" + "=".repeat(60));
        System.out.println("STARTING SHIFT ASSIGNMENT");
        System.out.println("=".repeat(60));
        
        // Reset schedule
        for (String day : DAYS) {
            for (String shift : SHIFTS) {
                schedule.get(day).get(shift).clear();
            }
        }
        
        for (String emp : employees) {
            daysWorked.put(emp, 0);
            employeeAssignedDays.put(emp, new HashSet<>());
        }
        
        // Phase 1: Assign based on preferences
        for (String employee : employees) {
            Map<String, List<String>> empPrefs = preferences.get(employee);
            
            for (String day : DAYS) {
                if (!empPrefs.containsKey(day)) continue;
                if (daysWorked.get(employee) >= MAX_DAYS_PER_WEEK) continue;
                if (employeeAssignedDays.get(employee).contains(day)) continue;
                
                boolean assigned = false;
                for (String shift : empPrefs.get(day)) {
                    schedule.get(day).get(shift).add(employee);
                    daysWorked.put(employee, daysWorked.get(employee) + 1);
                    employeeAssignedDays.get(employee).add(day);
                    System.out.println("Assigned " + employee + " to " + day + " - " + shift);
                    assigned = true;
                    break;
                }
                
                if (!assigned) {
                    System.out.println("Could not assign " + employee + " to " + day);
                }
            }
        }
        
        System.out.println("\n" + "-".repeat(60));
        System.out.println("FILLING UNDERSTAFFED SHIFTS");
        System.out.println("-".repeat(60));
        
        Random random = new Random();
        
        for (String day : DAYS) {
            for (String shift : SHIFTS) {
                int currentCount = schedule.get(day).get(shift).size();
                
                if (currentCount < MIN_EMPLOYEES_PER_SHIFT) {
                    int needed = MIN_EMPLOYEES_PER_SHIFT - currentCount;
                    System.out.println("\n" + day + " - " + shift + " needs " + needed + " more employee(s)");
                    
                    List<String> available = new ArrayList<>();
                    for (String emp : employees) {
                        if (daysWorked.get(emp) < MAX_DAYS_PER_WEEK && 
                            !employeeAssignedDays.get(emp).contains(day)) {
                            available.add(emp);
                        }
                    }
                    
                    if (!available.isEmpty()) {
                        Collections.shuffle(available, random);
                        int toAssign = Math.min(needed, available.size());
                        
                        for (int i = 0; i < toAssign; i++) {
                            String emp = available.get(i);
                            schedule.get(day).get(shift).add(emp);
                            daysWorked.put(emp, daysWorked.get(emp) + 1);
                            employeeAssignedDays.get(emp).add(day);
                            System.out.println("  Randomly assigned " + emp + " to " + day + " - " + shift);
                        }
                    } else {
                        System.out.println("  WARNING: No available employees for " + day + " - " + shift);
                    }
                }
            }
        }
        
        System.out.println("\n" + "=".repeat(60));
        System.out.println("SHIFT ASSIGNMENT COMPLETE");
        System.out.println("=".repeat(60) + "\n");
    }
    
    public void resolveConflicts() {
        System.out.println("\n" + "=".repeat(60));
        System.out.println("RESOLVING CONFLICTS");
        System.out.println("=".repeat(60));
        
        boolean conflictsFound = false;
        
        for (String day : DAYS) {
            List<String> employeesToday = new ArrayList<>();
            
            for (String shift : SHIFTS) {
                employeesToday.addAll(schedule.get(day).get(shift));
            }
            
            Set<String> seen = new HashSet<>();
            Set<String> duplicates = new HashSet<>();
            
            for (String emp : employeesToday) {
                if (seen.contains(emp)) {
                    duplicates.add(emp);
                }
                seen.add(emp);
            }
            
            if (!duplicates.isEmpty()) {
                conflictsFound = true;
                System.out.println("CONFLICT on " + day + ": " + 
                                 String.join(", ", duplicates) + " assigned to multiple shifts");
                
                for (String emp : duplicates) {
                    String keptShift = null;
                    
                    for (String shift : SHIFTS) {
                        List<String> shiftEmployees = schedule.get(day).get(shift);
                        
                        if (shiftEmployees.contains(emp)) {
                            if (keptShift == null) {
                                keptShift = shift;
                                System.out.println("  Keeping " + emp + " in " + shift);
                            } else {
                                shiftEmployees.remove(emp);
                                daysWorked.put(emp, daysWorked.get(emp) - 1);
                                System.out.println("  Removed " + emp + " from " + shift);
                            }
                        }
                    }
                }
            }
        }
        
        if (!conflictsFound) {
            System.out.println("No conflicts detected.");
        }
        
        System.out.println("=".repeat(60) + "\n");
    }
    
    public void displaySchedule() {
        System.out.println("\n" + "=".repeat(60));
        System.out.println("FINAL WEEKLY SCHEDULE");
        System.out.println("=".repeat(60) + "\n");
        
        for (String day : DAYS) {
            System.out.println("=".repeat(60));
            System.out.println(day.toUpperCase());
            System.out.println("=".repeat(60));
            
            for (String shift : SHIFTS) {
                List<String> shiftEmployees = schedule.get(day).get(shift);
                int empCount = shiftEmployees.size();
                
                if (!shiftEmployees.isEmpty()) {
                    String empList = String.join(", ", shiftEmployees);
                    String status = empCount >= MIN_EMPLOYEES_PER_SHIFT ? 
                                  "✓" : "⚠ (needs " + (MIN_EMPLOYEES_PER_SHIFT - empCount) + " more)";
                    
                    System.out.printf("  %-12s (%d employees) %s%n", shift, empCount, status);
                    System.out.println("               → " + empList);
                } else {
                    System.out.printf("  %-12s (0 employees) ⚠ UNDERSTAFFED%n", shift);
                }
            }
            
            System.out.println();
        }
        
        System.out.println("=".repeat(60));
        System.out.println("EMPLOYEE WORK SUMMARY");
        System.out.println("=".repeat(60));
        
        List<String> sortedEmployees = new ArrayList<>(employees);
        Collections.sort(sortedEmployees);
        
        for (String emp : sortedEmployees) {
            int days = daysWorked.get(emp);
            String status = days <= MAX_DAYS_PER_WEEK ? "✓" : "⚠ OVER LIMIT";
            System.out.printf("  %-20s %d days %s%n", emp, days, status);
        }
        
        System.out.println("=".repeat(60) + "\n");
    }
    
    public void getStatistics() {
        System.out.println("\n" + "=".repeat(60));
        System.out.println("SCHEDULING STATISTICS");
        System.out.println("=".repeat(60));
        
        int totalShifts = 0;
        int filledShifts = 0;
        int understaffedShifts = 0;
        
        for (String day : DAYS) {
            for (String shift : SHIFTS) {
                totalShifts++;
                int empCount = schedule.get(day).get(shift).size();
                
                if (empCount > 0) {
                    filledShifts++;
                }
                if (empCount < MIN_EMPLOYEES_PER_SHIFT) {
                    understaffedShifts++;
                }
            }
        }
        
        System.out.println("Total shifts: " + totalShifts);
        System.out.println("Filled shifts: " + filledShifts);
        System.out.println("Understaffed shifts: " + understaffedShifts);
        System.out.println("Total employees: " + employees.size());
        
        int totalDaysAssigned = 0;
        for (int days : daysWorked.values()) {
            totalDaysAssigned += days;
        }
        
        double avgDays = employees.isEmpty() ? 0.0 : (double) totalDaysAssigned / employees.size();
        System.out.printf("Average days per employee: %.2f%n", avgDays);
        
        System.out.println("=".repeat(60) + "\n");
    }
}

File: python/scheduler.py
import random
from collections import defaultdict
from typing import Dict, List, Set


class EmployeeScheduler:
    
    DAYS = ["Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday", "Sunday"]
    SHIFTS = ["Morning", "Afternoon", "Evening"]
    MAX_DAYS_PER_WEEK = 5
    MIN_EMPLOYEES_PER_SHIFT = 2
    
    def __init__(self):
        self.employees = []
        self.preferences = defaultdict(lambda: defaultdict(list))
        self.schedule = {day: {shift: [] for shift in self.SHIFTS} for day in self.DAYS}
        self.days_worked = defaultdict(int)
        self.employee_assigned_days = defaultdict(set)
    
    def add_employee(self, name: str):
        if name not in self.employees:
            self.employees.append(name)
            print(f"Added employee: {name}")
        else:
            print(f"Employee {name} already exists.")
    
    def add_preference(self, employee: str, day: str, shifts: List[str]):
        if employee not in self.employees:
            print(f"Error: Employee {employee} not found. Add them first.")
            return
        
        if day not in self.DAYS:
            print(f"Error: Invalid day {day}")
            return
        
        valid_shifts = [s for s in shifts if s in self.SHIFTS]
        if not valid_shifts:
            print(f"Error: No valid shifts provided")
            return
        
        self.preferences[employee][day] = valid_shifts
        print(f"Added preference for {employee} on {day}: {', '.join(valid_shifts)}")
    
    def assign_shifts(self):
        print("\n" + "="*60)
        print("STARTING SHIFT ASSIGNMENT")
        print("="*60)
        
        self.schedule = {day: {shift: [] for shift in self.SHIFTS} for day in self.DAYS}
        self.days_worked = defaultdict(int)
        self.employee_assigned_days = defaultdict(set)
        
        for employee in self.employees:
            for day in self.DAYS:
                if day in self.preferences[employee]:
                    if self.days_worked[employee] >= self.MAX_DAYS_PER_WEEK:
                        continue
                    
                    if day in self.employee_assigned_days[employee]:
                        continue
                    
                    assigned = False
                    for shift in self.preferences[employee][day]:
                        self.schedule[day][shift].append(employee)
                        self.days_worked[employee] += 1
                        self.employee_assigned_days[employee].add(day)
                        print(f"Assigned {employee} to {day} - {shift}")
                        assigned = True
                        break
                    
                    if not assigned:
                        print(f"Could not assign {employee} to {day}")
        
        print("\n" + "-"*60)
        print("FILLING UNDERSTAFFED SHIFTS")
        print("-"*60)
        
        for day in self.DAYS:
            for shift in self.SHIFTS:
                current_count = len(self.schedule[day][shift])
                
                if current_count < self.MIN_EMPLOYEES_PER_SHIFT:
                    needed = self.MIN_EMPLOYEES_PER_SHIFT - current_count
                    print(f"\n{day} - {shift} needs {needed} more employee(s)")
                    
                    available = []
                    for emp in self.employees:
                        if (self.days_worked[emp] < self.MAX_DAYS_PER_WEEK and 
                            day not in self.employee_assigned_days[emp]):
                            available.append(emp)
                    
                    if available:
                        random.shuffle(available)
                        for i in range(min(needed, len(available))):
                            emp = available[i]
                            self.schedule[day][shift].append(emp)
                            self.days_worked[emp] += 1
                            self.employee_assigned_days[emp].add(day)
                            print(f"  Randomly assigned {emp} to {day} - {shift}")
                    else:
                        print(f"  WARNING: No available employees for {day} - {shift}")
        
        print("\n" + "="*60)
        print("SHIFT ASSIGNMENT COMPLETE")
        print("="*60 + "\n")
    
    def resolve_conflicts(self):
        print("\n" + "="*60)
        print("RESOLVING CONFLICTS")
        print("="*60)
        
        conflicts_found = False
        
        for day in self.DAYS:
            employees_today = []
            for shift in self.SHIFTS:
                employees_today.extend(self.schedule[day][shift])
            
            seen = set()
            duplicates = set()
            for emp in employees_today:
                if emp in seen:
                    duplicates.add(emp)
                seen.add(emp)
            
            if duplicates:
                conflicts_found = True
                print(f"CONFLICT on {day}: {', '.join(duplicates)} assigned to multiple shifts")
                
                for emp in duplicates:
                    kept_shift = None
                    for shift in self.SHIFTS:
                        if emp in self.schedule[day][shift]:
                            if kept_shift is None:
                                kept_shift = shift
                                print(f"  Keeping {emp} in {shift}")
                            else:
                                self.schedule[day][shift].remove(emp)
                                self.days_worked[emp] -= 1
                                print(f"  Removed {emp} from {shift}")
        
        if not conflicts_found:
            print("No conflicts detected.")
        
        print("="*60 + "\n")
    
    def display_schedule(self):
        print("\n" + "="*60)
        print("FINAL WEEKLY SCHEDULE")
        print("="*60 + "\n")
        
        for day in self.DAYS:
            print("="*60)
            print(day.upper())
            print("="*60)
            
            for shift in self.SHIFTS:
                employees = self.schedule[day][shift]
                emp_count = len(employees)
                
                if employees:
                    emp_list = ", ".join(employees)
                    status = "✓" if emp_count >= self.MIN_EMPLOYEES_PER_SHIFT else f"⚠ (needs {self.MIN_EMPLOYEES_PER_SHIFT - emp_count} more)"
                    print(f"  {shift:12} ({emp_count} employees) {status}")
                    print(f"               → {emp_list}")
                else:
                    print(f"  {shift:12} (0 employees) ⚠ UNDERSTAFFED")
            
            print()
        
        print("="*60)
        print("EMPLOYEE WORK SUMMARY")
        print("="*60)
        
        for emp in sorted(self.employees):
            days = self.days_worked[emp]
            status = "✓" if days <= self.MAX_DAYS_PER_WEEK else "⚠ OVER LIMIT"
            print(f"  {emp:20} {days} days {status}")
        
        print("="*60 + "\n")
    
    def get_statistics(self):
        print("\n" + "="*60)
        print("SCHEDULING STATISTICS")
        print("="*60)
        
        total_shifts = 0
        filled_shifts = 0
        understaffed_shifts = 0
        
        for day in self.DAYS:
            for shift in self.SHIFTS:
                total_shifts += 1
                emp_count = len(self.schedule[day][shift])
                if emp_count > 0:
                    filled_shifts += 1
                if emp_count < self.MIN_EMPLOYEES_PER_SHIFT:
                    understaffed_shifts += 1
        
        print(f"Total shifts: {total_shifts}")
        print(f"Filled shifts: {filled_shifts}")
        print(f"Understaffed shifts: {understaffed_shifts}")
        print(f"Total employees: {len(self.employees)}")
        
        total_days_assigned = sum(self.days_worked.values())
        avg_days = total_days_assigned / len(self.employees) if self.employees else 0
        print(f"Average days per employee: {avg_days:.2f}")
        
        print("="*60 + "\n")
